FABulous/timing_model/FABulous_timing_model.py

```diff
@@ -0,0 +1,389 @@
+        
+        print("Extracting switch matrix information...")
+        self.switch_matrix_hier_path = self.hdlnx_tm_synth.find_instance_paths_by_regex(r".*_switch_matrix$")
+        self.switch_matrix_module_name = self.hdlnx_tm_synth.find_verilog_modules_regex(r"^[^/]*_switch_matrix$")
+        
+        if self.config["super_tile_type"] is not None and self.config["super_tile_type"] in ["top", "bot"]:
+            self.switch_matrix_hier_path = [p for p in self.switch_matrix_hier_path 
+                if f"{self.config['tile_name']}_{self.config['super_tile_type']}_" in p
+            ]
+            self.switch_matrix_module_name = [m for m in self.switch_matrix_module_name 
+                if f"{self.config['tile_name']}_{self.config['super_tile_type']}_" in m
+            ]
+            if len(self.switch_matrix_hier_path) == 0 or len(self.switch_matrix_module_name) == 0:
+                raise ValueError(f"No switch matrix instance or module found for super_tile_type "
+                                 f"{self.config['super_tile_type']}")
+            if len(self.switch_matrix_hier_path) > 1 or len(self.switch_matrix_module_name) > 1:
+                raise ValueError(f"Multiple switch matrix instances or modules found for super_tile_type "
+                                 f"{self.config['super_tile_type']}")
+            
+            self.switch_matrix_hier_path = self.switch_matrix_hier_path[0]
+            self.switch_matrix_module_name = self.switch_matrix_module_name[0]
+            
+            print(f"Using super_tile_type {self.config['super_tile_type']} from config.")
+            print(f"Using switch matrix instance: {self.switch_matrix_hier_path}, module: {self.switch_matrix_module_name}")
+        else:
+            if len(self.switch_matrix_hier_path) == 0 or len(self.switch_matrix_module_name) == 0:
+                raise ValueError("No switch matrix instance or module found.")
+            if len(self.switch_matrix_hier_path) > 1 or len(self.switch_matrix_module_name) > 1:
+                raise ValueError("Multiple switch matrix instances or modules found. Please specify super_tile_type in config.")
+            
+            self.switch_matrix_hier_path = self.switch_matrix_hier_path[0]
+            self.switch_matrix_module_name = self.switch_matrix_module_name[0]
+            
+            print(f"Using switch matrix instance: {self.switch_matrix_hier_path}, module: {self.switch_matrix_module_name}")
+        
+        print("Loading internal PIPs...")
+        self.internal_pips_grouped_by_inst = self.hdlnx_tm_synth.get_module_instance_nets(self.switch_matrix_module_name)
+        self.internal_pips = self.hdlnx_tm_synth.get_instance_pins(self.switch_matrix_hier_path)
+        print("FABulous Timing Model initialized.")
+        
+        ### Definitions:
+        
+        self.mode = self.config["mode"]
+    
+    
+    def _add_config_keys(self, new_keys: dict, msg: str = ""):
```

Comment:
Since your config has sa pecial function, you  probably want to make the config into a class itself. If you want to keep the dict syntax, you can also override `__getitems__` and `__setitems__`.